cmd/shimdiag/shimdiag.go

```diff
@@ -0,0 +1,102 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/Microsoft/go-winio"
+	"github.com/containerd/ttrpc"
+	"github.com/urfave/cli"
+	"golang.org/x/sys/windows"
+)
+
+const (
+	shimPrefix = `\\.\pipe\ProtectedPrefix\Administrators\containerd-shim-`
+	shimSuffix = `-pipe`
+)
+
+func main() {
+	app := cli.NewApp()
+	app.Name = "shimdiag"
+	app.Usage = "runhcs shim diagnostic tool"
+	app.Commands = []cli.Command{
+		listCommand,
+		execCommand,
+	}
+	if err := app.Run(os.Args); err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(1)
+	}
+}
+
+func findPipes(pattern string) ([]string, error) {
+	path := `\\.\pipe\*`
+	path16, err := windows.UTF16FromString(path)
```

Comment:
Can you not use `shimPrefix+"*"`?

Reply:
No because to enumerate pipes, you open \\.\pipe and read the list, but FindFirstFile, when it sees \\.\pipe\ProtectedPrefix\Foo\Bar* will try to open \\.\pipe\ProtectedPrefix\Foo, which will fail.

This is possible to work around at the NT layer, but that's too much of a pain.